File: python/image_decode_task_status_client.py
import struct
import socket
import http.client

import server_utils

class TaskStatusClient:
    def __init__(self, ip, port):
        self.ip = ip
        self.port = port

    def get_task_status(self):
        raise NotImplementedError()

class TaskStatusTcpClient(TaskStatusClient):
    def __init__(self, ip, port):
        super().__init__(ip, port)

    def get_task_status(self):
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.settimeout(5)
                s.connect((self.ip, self.port))
                len_bytes = bytearray()
                while len(len_bytes) < 8:
                    len_bytes += s.recv(8 - len(len_bytes))
                task_byte_len = struct.unpack('<Q', len_bytes)[0]
                task_bytes = bytearray()
                if task_byte_len:
                    while len(task_bytes) < task_byte_len:
                        task_bytes += s.recv(task_byte_len - len(task_bytes))
                return task_bytes
        except Exception as e:
            return None

class TaskStatusHttpClient(TaskStatusClient):
    def __init__(self, ip, port):
        super().__init__(ip, port)

    def get_task_status(self):
        try:
            conn = http.client.HTTPConnection(self.ip, self.port)
            conn.request('GET', '/')
            resp = conn.getresponse()
            task_bytes = resp.read(int(resp.getheader('Content-Length')))
            return task_bytes
        except Exception as e:
            return None

server_type_to_client_mapping = {
    server_utils.ServerType.TCP: TaskStatusTcpClient,
    server_utils.ServerType.HTTP: TaskStatusHttpClient,
    }

def create_task_status_client(server_type, ip, port):
    return server_type_to_client_mapping[server_type](ip, port)

File: src/image_codec/image_decode_task_status_client.cpp
#include <boost/asio.hpp>
#include <boost/beast.hpp>

#include "image_decode_task_status_client.h"

TaskStatusClient::TaskStatusClient(const std::string& ip, int port) : m_ip(ip), m_port(port) {
}

Bytes TaskStatusTcpClient::GetTaskStatus() {
    try {
        boost::asio::io_context io_context;
        boost::asio::ip::tcp::socket socket(io_context);
        boost::asio::ip::tcp::endpoint endpoint(boost::asio::ip::make_address(GetIp()), GetPort());
        socket.connect(endpoint);
        uint64_t task_byte_len = 0;
        uint64_t done1 = 0;
        while (done1 < sizeof(uint64_t)) {
            done1 += boost::asio::read(socket, boost::asio::buffer(reinterpret_cast<char*>(&task_byte_len) + done1, sizeof(uint64_t) - done1));
        }
        Bytes task_bytes(task_byte_len);
        if (task_byte_len) {
            uint64_t done2 = 0;
            while (done2 < task_byte_len) {
                done2 += boost::asio::read(socket, boost::asio::buffer(task_bytes.data() + done2, task_byte_len - done2));
            }
        }
        return task_bytes;
    }
    catch (std::exception& e) {
        //std::cerr << e.what() << "\n";
        return Bytes();
    }
}

Bytes TaskStatusHttpClient::GetTaskStatus() {
    try {
        boost::asio::io_context ioc;
        boost::asio::ip::tcp::socket socket(ioc);
        boost::asio::ip::tcp::endpoint endpoint(boost::asio::ip::make_address(GetIp()), GetPort());
        socket.connect(endpoint);
        boost::beast::http::request<boost::beast::http::string_body> req(boost::beast::http::verb::get, "/", 10);
        req.set(boost::beast::http::field::user_agent, BOOST_BEAST_VERSION_STRING);
        boost::beast::http::write(socket, req);
        boost::beast::flat_buffer buffer;
        boost::beast::http::response<boost::beast::http::vector_body<Byte>> res;
        boost::beast::http::read(socket, buffer, res);
        return res.body();
    }
    catch (std::exception& e) {
        //std::cerr << e.what() << "\n";
        return Bytes();
    }
}

std::unique_ptr<TaskStatusClient> create_task_status_client(ServerType server_type, const std::string& ip, int port) {
    std::unique_ptr<TaskStatusClient> task_status_client;
    if (server_type == ServerType::TCP) {
        task_status_client = std::make_unique<TaskStatusTcpClient>(ip, port);
    } else if (server_type == ServerType::HTTP) {
        task_status_client = std::make_unique<TaskStatusHttpClient>(ip, port);
    } else {
        throw std::invalid_argument("invalid server type '" + std::to_string(static_cast<int>(server_type)) + "'");
    }
    return task_status_client;
}
